src/interchange/src/avro.rs

```diff
@@ -2068,20 +2010,54 @@ impl fmt::Debug for Encoder {
 }
 
 impl Encoder {
-    pub fn new(desc: RelationDesc, include_transaction: bool) -> Self {
+    fn key_filter(&self) -> Option<&[bool]> {
+        self.key_schema_and_filter
+            .as_ref()
+            .map(|(_, filter)| filter.as_slice())
+    }
+    pub fn new(
+        desc: RelationDesc,
+        include_transaction: bool,
+        key_indices: Option<Vec<usize>>,
+    ) -> Self {
         let columns = column_names_and_types(desc);
         let writer_schema = build_schema(&columns, include_transaction);
+        let key_schema_and_filter = key_indices.map(|key_indices| {
+            let key_columns = key_indices
+                .iter()
+                .map(|&key_idx| columns[key_idx].clone())
+                .collect::<Vec<_>>();
+            let row_schema = build_row_schema_json(&key_columns, "row");
+            let mut filter = std::iter::repeat(false)
```

Comment:
I think this can be simplified

```
diff --git a/src/interchange/src/avro.rs b/src/interchange/src/avro.rs
index 4b93b6aa8..e95c3e4cb 100644
--- a/src/interchange/src/avro.rs
+++ b/src/interchange/src/avro.rs
@@ -2028,9 +2028,7 @@ impl Encoder {
                 .map(|&key_idx| columns[key_idx].clone())
                 .collect::<Vec<_>>();
             let row_schema = build_row_schema_json(&key_columns, "row");
-            let mut filter = std::iter::repeat(false)
-                .take(columns.len())
-                .collect::<Vec<_>>();
+            let mut filter = vec![false; columns.len()];
```